Define desired kubernetes object names

The builder needs to use suitably unique names for the kubernetes "stuff" required for running chaincode

The run binary has access to the unique ID associated with the chaincode package, which is in the form of `label:hash`, the address of the peer, which is in the form of `host:port`, and the local mspid of the peer.

The builder can also use environment variables which are configured in core.yaml, for example `CORE_PEER_ID` can be propagated to the builder, and currently is required. Using the peer address would reduce the configuration required and may be preferable.

**Important:** whatever combination is used needs to be sanitised and truncated to give a [valid kubernetes object name](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/), and the naming rules differ between objects. For example the peer ID can be relatively long, e.g. `http://peer0.org1.example.com/` in the nano test network, and the same for the chaincode label, plus the chaincode ID includes a full hash. All these will need to be truncated and have any invalid characters removed or replaced.

This should fix the following problems (which need tests!)

- secret name...
```
unable to create kubernetes secret for chaincode ID CHAINCODE_ID: Secret "cc-mspid-core_peer_id_abcdefghijklmnopqrstuvwxyz_0123456789chai" is invalid: metadata.name: Invalid value: "cc-mspid-core_peer_id_abcdefghijklmnopqrstuvwxyz_0123456789chai": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')
```
- util.mangleName...
```
panic: runtime error: slice bounds out of range [:63] with length 33
```

--- a/cmd/run/main_test.go
+++ b/cmd/run/main_test.go
@@ -59,7 +59,7 @@ var _ = Describe("Main", func() {
 			).Should(gbytes.Say(`run \[\d+\] DEBUG: FABRIC_K8S_BUILDER_SERVICE_ACCOUNT=chaincode`))
 			Eventually(
 				session.Err,
-			).Should(gbytes.Say(`run \[\d+\]: Running chaincode ID CHAINCODE_ID in kubernetes pod chaincode/cc-mspid-core-peer-id-abcdefghijklmnopqrstuvwxyz-0123456789chai`))
+			).Should(gbytes.Say(`run \[\d+\]: Running chaincode ID CHAINCODE_ID in kubernetes pod chaincode/cc-m4eml38l6lv2ost7v1i3q6698a5p4gu1l3lnflb4boi7jnle6lt0`))
 
 			pipe := script.Exec(
 				"kubectl wait --for=condition=ready pod --timeout=120s --namespace=chaincode -l fabric-builder-k8s-peerid=core-peer-id-abcdefghijklmnopqrstuvwxyz-0123456789",

--- a/internal/util/k8s.go
+++ b/internal/util/k8s.go
@@ -4,10 +4,11 @@ package util
 
 import (
 	"context"
+	"crypto/sha256"
+	"encoding/base32"
 	"encoding/base64"
 	"fmt"
 	"os"
-	"regexp"
 	"strings"
 	"time"
 
@@ -36,8 +37,6 @@ const (
 	TLSClientCertFile     string = "/etc/hyperledger/fabric/client_pem.crt"
 	TLSClientRootCertFile string = "/etc/hyperledger/fabric/peer.crt"
 )
-
-var mangledRegExp = regexp.MustCompile("[^a-zA-Z0-9-_.]")
 
 func waitForPod(
 	ctx context.Context,
@@ -303,7 +302,7 @@ func getChaincodePodObject(
 					Name: "certs",
 					VolumeSource: apiv1.VolumeSource{
 						Secret: &apiv1.SecretVolumeSource{
-							SecretName: getSecretName(
+							SecretName: GetValidName(
 								chaincodeData.MspID,
 								peerID,
 								chaincodeData.ChaincodeID,
@@ -320,7 +319,7 @@ func getChaincodeSecretApplyConfiguration(
 	namespace, peerID string,
 	chaincodeData *ChaincodeJSON,
 ) *applycorev1.SecretApplyConfiguration {
-	name := getSecretName(chaincodeData.MspID, peerID, chaincodeData.ChaincodeID)
+	name := GetValidName(chaincodeData.MspID, peerID, chaincodeData.ChaincodeID)
 
 	annotations := map[string]string{
 		"fabric-builder-k8s-ccid": chaincodeData.ChaincodeID,
@@ -349,10 +348,6 @@ func getChaincodeSecretApplyConfiguration(
 		WithLabels(labels).
 		WithStringData(data).
 		WithType(apiv1.SecretTypeOpaque)
-}
-
-func getPodName(mspID, peerID, chaincodeID string) string {
-	return mangleName("cc-" + mspID + "-" + peerID + chaincodeID)
 }
 
 func ApplyChaincodeSecrets(
@@ -443,7 +438,7 @@ func CreateChaincodePod(
 	chaincodeData *ChaincodeJSON,
 	imageData *ImageJSON,
 ) (*apiv1.Pod, error) {
-	podName := getPodName(chaincodeData.MspID, peerID, chaincodeData.ChaincodeID)
+	podName := GetValidName(chaincodeData.MspID, peerID, chaincodeData.ChaincodeID)
 	podDefinition := getChaincodePodObject(
 		imageData,
 		namespace,
@@ -492,11 +487,13 @@ func CreateChaincodePod(
 	return pod, nil
 }
 
-func getSecretName(mspID, peerID, chaincodeID string) string {
-	return mangleName("cc-" + mspID + "-" + peerID + chaincodeID)
-}
+// GetValidName returns a valid RFC 1035 label name.
+func GetValidName(mspID, peerID, chaincodeID string) string {
+	qualifiedChaincodeID := mspID + ":" + peerID + ":" + chaincodeID
+	h := sha256.New()
+	h.Write([]byte(qualifiedChaincodeID))
+	sum := h.Sum(nil)
+	encodedChaincodeID := base32.HexEncoding.WithPadding(base32.NoPadding).EncodeToString(sum)
 
-func mangleName(name string) string {
-	// TODO need sensible unique naming scheme for deployments and secrets!
-	return strings.ToLower(mangledRegExp.ReplaceAllString(name, "-")[:63])
+	return "cc-" + strings.ToLower(encodedChaincodeID)
 }

--- a/internal/util/k8s_test.go
+++ b/internal/util/k8s_test.go
@@ -1,0 +1,42 @@
+package util_test
+
+import (
+	"github.com/hyperledger-labs/fabric-builder-k8s/internal/util"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+)
+
+var _ = Describe("K8s", func() {
+	Describe("GetValidName", func() {
+		It("should return a string with a maximum of 63 characters", func() {
+			name := util.GetValidName("CongaCongaCongaCongaCongaCongaCongaCongaCongaCongaCongaCongaOrgMsp", "CongaCongaCongaCongaCongaCongaCongaCongaCongaCongaCongaCongaOrgPeer0", "fabfabfabfabcar:cffa266294278404e5071cb91150d550dc0bf855149908a170b1169d6160004b")
+			Expect(len(name)).To(BeNumerically("<=", 63))
+		})
+
+		It("should return a string which starts with an alphabetic character", func() {
+			name := util.GetValidName("GreenCongaOrg", "GreenCongaOrgPeer0", "fabcar:cffa266294278404e5071cb91150d550dc0bf855149908a170b1169d6160004b")
+			Expect(name).To(MatchRegexp("^[a-z]"))
+		})
+
+		It("should return a string which ends with an alphanumeric character", func() {
+			name := util.GetValidName("BlueCongaOrg", "BlueCongaOrgPeer0", "fabcar:cffa266294278404e5071cb91150d550dc0bf855149908a170b1169d6160004b")
+			Expect(name).To(MatchRegexp("[a-z0-9]$"))
+		})
+
+		It("should return a string which only contains lowercase alphanumeric characters or '-'", func() {
+			name := util.GetValidName("BlueCongaOrg", "BlueCongaOrgPeer0", "fabcar:cffa266294278404e5071cb91150d550dc0bf855149908a170b1169d6160004b")
+			Expect(name).To(MatchRegexp("^(?:[a-z0-9]|-)+$"))
+		})
+
+		It("should return different names for different input", func() {
+			name1 := util.GetValidName("GreenCongaOrg", "GreenCongaOrgPeer0", "fabcar:cffa266294278404e5071cb91150d550dc0bf855149908a170b1169d6160004b")
+			name2 := util.GetValidName("BlueCongaOrg", "BlueCongaOrgPeer0", "fabcar:cffa266294278404e5071cb91150d550dc0bf855149908a170b1169d6160004b")
+			Expect(name1).NotTo(Equal(name2))
+		})
+
+		It("should return deterministic names", func() {
+			name := util.GetValidName("CongaOrg", "CongaOrgPeer0", "fabcar:cffa266294278404e5071cb91150d550dc0bf855149908a170b1169d6160004b")
+			Expect(name).To(Equal("cc-ocqvh9ir0mi0ef6urh12f3l0dar6csdmtjfhgbfvdp2d22u109r0"))
+		})
+	})
+})

--- a/internal/util/util_suite_test.go
+++ b/internal/util/util_suite_test.go
@@ -1,0 +1,13 @@
+package util_test
+
+import (
+	"testing"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+)
+
+func TestUtil(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "Util Suite")
+}

